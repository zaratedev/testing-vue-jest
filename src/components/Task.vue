<template>
    <div 
        class="list-group-item d-flex justify-content-between align-items-center"
        :class="{ disabled: task.done }"
    >
        <div>
            {{Â task.name }}<input class="ml-3" v-model="isChecked" type="checkbox">
            <router-link :to="{ path: '/task', params: { task } }"></router-link>
        </div>
        <button id="delete" type="button" class="btn btn-sm btn-danger" @click="$emit('delete')">
            Delete
        </button>
    </div>
</template>

<script>
import { mapActions } from 'vuex';

export default {
    name: 'Task',

    props: {
        task: Object
    },

    computed: {
        isChecked: {
            get() {
                if (! this.task) {
                    return false;
                }

                return this.task.done;
            },
            set(value) {
                if (value) {
                    this.completeTask(this.task);
                } else {
                    this.uncompleteTask(this.task);
                }
            }
        }
    },

    methods: {
        ...mapActions(['completeTask', 'uncompleteTask'])
    }
}
</script>
