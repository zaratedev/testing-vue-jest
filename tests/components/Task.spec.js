import { mount, shallowMount, RouterLinkStub, config } from '@vue/test-utils';
import Task from '@/components/Task';

config.stubs['router-link'] = '<div />';
describe('Component Task', () => {
  let mocks;

  beforeEach(() => {
    mocks = {
      $store: {
        dispatch: jest.fn()
      }
    }
  });

  test('it renders task prop', () => {
    const wrapper = mount(Task, {
      propsData: {
        task: { name: 'My new Task' }
      }
    });
    expect(wrapper.text()).toContain('My new Task');
  });

  test('it renders task prop', () => {
    const wrapper = mount(Task);
    wrapper.setProps({task: 'My new Task'});
    expect(wrapper.text()).toContain('My new Task');
  });

  test('it emits delete event when delete button is clicked', () => {
    const wrapper = mount(Task);
    const button = wrapper.find('#delete');
    button.trigger('click');
    expect(wrapper.emitted().delete).toBeTruthy();
  });

  test('it renders default slot', () => {
    const wrapper = shallowMount(Task, {
      slots: {
        default: '<div id="close"></div>'
      }
    });
    expect(wrapper.contains('#close')).toBe(true);
  });

  test('it renders the TaskView router-link ', () => {
    const wrapper = shallowMount(Task, {
      stubs: {
        'router-link': RouterLinkStub
      },
      propsData: { task: { name: 'My Task' } },
    });

    expect(wrapper.find(RouterLinkStub).props().to).toEqual({
      path: '/task',
      params: {
        task: { name: 'My Task' },
      }
    });
  });

  test('checkbox is checked if task.done is true', () => {
    const wrapper = shallowMount(Task, {
      propsData: {
        task: { name: 'My Task', done: true },
      }
    });

    const checkbox = wrapper.find('input[type="checkbox"]');

    expect(checkbox.element.checked).toBeTruthy();
  });

  test('checkbox is not checked if task.done is false', () => {
    const wrapper = shallowMount(Task, {
      propsData: {
        task: { name: 'My Task', done: false },
      }
    });

    const checkbox = wrapper.find('input[type="checkbox"]');

    expect(checkbox.element.checked).toBeFalsy();
  });

  test('it calls to completeTask action when checkbox is clicked and is false', () => {
    const wrapper = shallowMount(Task, {
      mocks,
      propsData: {
        task: { name: 'My task' }
      }
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    checkbox.element.checked = false;
    checkbox.trigger('click');

    expect(mocks.$store.dispatch).toHaveBeenCalledTimes(1);
    expect(mocks.$store.dispatch.mock.calls[0][0]).toEqual('completeTask');
    expect(mocks.$store.dispatch.mock.calls[0][1]).toEqual({ name: 'My task' });
  });

  test('it calls to uncompleteTask action when checkbox is clicked and is true', () => {
    const wrapper = shallowMount(Task, {
      mocks,
      propsData: {
        task: { name: 'My task' }
      }
    });
    const checkbox = wrapper.find('input[type="checkbox"]');
    checkbox.element.checked = true;
    checkbox.trigger('click');

    expect(mocks.$store.dispatch).toHaveBeenCalledTimes(1);
    expect(mocks.$store.dispatch.mock.calls[0][0]).toEqual('uncompleteTask');
    expect(mocks.$store.dispatch.mock.calls[0][1]).toEqual({ name: 'My task' });
  });
});
