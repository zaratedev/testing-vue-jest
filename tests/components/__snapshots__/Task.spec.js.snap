// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Task renders correctly 1`] = `
"<div class=\\"list-group-item d-flex justify-content-between align-items-center\\"><div>
        My Task<input type=\\"checkbox\\" class=\\"ml-3\\"> <div to=\\"[object Object]\\"></div></div> <button id=\\"delete\\" type=\\"button\\" class=\\"btn btn-sm btn-danger\\">
        Delete
    </button></div>"
`;
